backend/server/src/derivation.rs

```diff
@@ -0,0 +1,142 @@
+//! Derivation build data model.
+
+use std::num::NonZero;
+
+/// A derivation id combined with a counter that keeps track of the number of build attempts for
+/// this derivation.
+pub struct DrvBuildId {
+    /// The derivation that is attempted to be build.
+    pub derivation: DrvId,
+
+    /// Denotes the number of times this derivation has been attempted to be build.
+    ///
+    /// Note that once the build outcome of a derivation has been determined, there is no point in
+    /// trying to build the same derivation again. If it failed once, it will always fail.
+    ///
+    /// This counter is intended for cases in which the derivation build was interrupted due to
+    /// external factors (see [DrvBuildState::Interrupted]). In these situations it may make sense
+    /// to reattempt the build (depending on the interruption kind).
+    pub build_attempt: NonZero<u32>,
```

Comment:
I think we can de-normalize this a bit, and just add this to the `DrvBuildMetadata` struct, and remove the need for DrvBuildId altogether.

Reply:
We also need it for `DrvBuildEvent` though. My thinking here was to have a single place for writing the documentation. 

Same for `DrvId`, which can additionally ensure only valid identifiers are stored in the database (its `String` is not public, so it can not be directly constructed from outside the module).

I am a big fan of utilizing Rust's typesystem this way (see also "parse don't validate").

If you however feel that this is to much, I am open to adjusting it as necessary.

Reply:
Oh, right. `DrvBuildId` and `DrvId` are two different things. Ignore the second part of my reply xD

Reply:
> We also need it for DrvBuildEvent though. My thinking here was to have a single place for writing the documentation.

I agree, buildevent makes sense

Reply:
> We also need it for DrvBuildEvent though. My thinking here was to have a single place for writing the documentation.

I agree